adminHome.js does not URL encode form data before passing it on as GET params #6397

thenaesh opened the pull request:
Fixes #6397 
<!--
Please follow the naming conventions in the "guidelines for contributing" link above or the pull request may be rejected.

Also, do consider outlining the solution taken as doing so will likely help in the reviewing process
(e.g., when the pull request involves non-trivial changes)
-->

thenaesh commented:
@whipermr5 Unfortunately, encoding the whole email address causes the email address to fail the email address validation (mainly because `@` gets encoded). I'm looking into fixing the validation code so that this becomes a non-issue.

whipermr5 commented:
It's not the email address validation that's failing, but the course ID validation. Upon adding of a new instructor, a demo course is generated based on the instructor email, but the generated course ID imports the `+` from the email address (see #6398). I've fixed #6398 locally and indeed, encoding the entire email address works perfectly.

thenaesh commented:
Oh ok, that makes sense now. Thanks!

thenaesh commented:
Ready to review.

whipermr5 commented on src/main/webapp/js/adminHome.js:
```diff
@@ -30,6 +30,41 @@ function createRowForResultTable(shortName, name, email, institution, isSuccess,
+ * @returns the encoded query string
+ */
+function encodeQueryString(queryString) {
+    return queryString.split('&')
```
What if the user input contains `&`? I feel the encoding should happen before the querystring is constructed

whipermr5 replied:
What if the user input contains `&`? The data should be encoded before constructing the query string. Or you could use [`jQuery.param`](http://api.jquery.com/jQuery.param/) to do both together

thenaesh replied:
Oh yea, good point.

whipermr5 commented on src/main/webapp/js/adminHome.js:
```diff
@@ -145,12 +180,12 @@ function addInstructorFromSecondFormByAjax() {
                             + '|' + $('#instructorInstitution').val();
-    instructorDetailsList = [instructorDetails];
+    instructorDetailsList = [encodeInstructorDetails(instructorDetails)];
     var params = 'instructorshortname=' + $('#instructorShortName').val()
```
I.e. `var params = 'instructorshortname=' + encodeURIComponent($('#instructorShortName').val())` etc

thenaesh replied:
Just had a look, `$.param` looks like it actually encodes the string properly, so all the separate encoding functions are unnecessary. I'll modify that part to use a query object then.

whipermr5 replied:
Yup 👍👍

thenaesh commented:
@whipermr5 I've made the changes you suggested. I removed the extra functions at the top and just used `encodeURIComponent` directly. Also, `$.param` makes the code look nicer I think, no more + signs all over the place. xD

If you have any other suggestions, do let me know.

whipermr5 commented:
Nice! Looks good to me :)

thenaesh commented:
Two issues:
1) I think it is better to wait for #6408 to be merged, since the error messages I use in the test will change.
2) I just discovered that this PR will open up a latent vulnerability, which I'm trying to fix first.

thenaesh commented:
@wkurniawan07 Done! Ready for review.

whipermr5 commented:
Looks good! Without your changes, the updated test case would fail as `+` would be interpreted as a space.

wkurniawan07 commented on src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java:
```diff
@@ -74,14 +74,24 @@ private void testCreateInstructorAction() throws Exception {
+        BackDoor.deleteAccount(TestProperties.TEST_INSTRUCTOR_ACCOUNT);
+        BackDoor.deleteCourse(demoCourseId);
+        BackDoor.deleteInstructor(demoCourseId, instructor.email);
+        homePage.createInstructorByInstructorDetailsSingleLineForm(instructorDetails);
```
What's the reason for this new sequence of action? From what I understand, the two `assertEquals` below can be moved to the pre-existing test sequence immediately after this one.

thenaesh replied:
Oh yea you're right. I initially used a different set of instructor details for this test alone, so I had to do this. Then I realised that it wasn't necessary to use a different set of details and removed those, but forgot to remove this. I'll fix it. Thanks for pointing it out.

thenaesh commented:
@wkurniawan07 I've removed that duplicate code.

wkurniawan07 commented on src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java:
```diff
@@ -74,19 +74,25 @@ private void testCreateInstructorAction() throws Exception {
                                  + instructor.name + " | " + instructor.email + " | " + institute;
         
-        ______TS("action fail & success: add multiple instructors");
+        ______TS("action success: instructor details are sent across to the backend properly");
```
It seems like there are unnecessary changes now, like this.

wkurniawan07 commented on src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java:
```diff
@@ -74,19 +74,25 @@ private void testCreateInstructorAction() throws Exception {
+        assertEquals(instructor.getName(), instructorInBackend.getName());
+        assertEquals(instructor.getEmail(), instructorInBackend.getEmail());
+        
+        ______TS("action fail & success: add multiple instructors");
```
And this as well. This is why my previous suggestion was to move the two `assertEquals`. To be more precise, to between the `assertEquals("Instructor AHPUiT Instructor+++...` and `homePage.clearInstructorDetailsSingleLineForm();`.

thenaesh commented:
@wkurniawan07 Moved the `assert`s to between the existing test sequence as per your recommendations. I removed the test segment I added earlier though, as it looks weird in that location. xD

damithc commented on src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java:
```diff
@@ -74,10 +74,10 @@ private void testCreateInstructorAction() throws Exception {
         String shortName = "Instrúctör";
-        instructor.name = "AHPUiT Instrúctör";
-        instructor.email = "AHPUiT.instr1!@gmail.tmt";
+        instructor.name = "AHPUiT Instrúctör+++";
```
Can this made more realistic? e.g. `AHPUiT Instrúctör WithPlusInEmail` Names don't have `+` in them. Unrealistic data makes the test outputs look unrealistic, making it hard to detect rubbish output caused by bugs.

thenaesh commented:
@damithc Changes made.

damithc commented on src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java:
```diff
@@ -74,10 +74,10 @@ private void testCreateInstructorAction() throws Exception {
         String shortName = "Instrúctör";
-        instructor.name = "AHPUiT Instrúctör";
-        instructor.email = "AHPUiT.instr1!@gmail.tmt";
+        instructor.name = "AHPUiT Instrúctör WithNameInEmail";
```
`WithNameInEmail` doesn't indicate that I should expect weird chars in email address of this guy. Furthermore, most test data follow the pattern that user name is partially/fully included in the email address, so there is no reason to mention it in just this test case. Either omit it or use a more meaningful 'name'?

thenaesh replied:
Oops, typo.
